lib/class-wp-rest-url-details-controller.php

```diff
@@ -208,13 +214,92 @@ private function get_remote_url( $url ) {
 	 * @return string the title tag contents (maybe empty).
 	 */
 	private function get_title( $html ) {
-		preg_match( '|<title>([^<]*?)</title>|is', $html, $match_title );
+		preg_match( '|<\s*title[^>]*>(.*?)<\s*/\s*title>|is', $html, $match_title );
 
-		$title = isset( $match_title[1] ) ? trim( $match_title[1] ) : '';
+		$title = isset( $match_title[1] ) && is_string( $match_title[1] ) ? trim( $match_title[1] ) : '';
 
 		return $title;
 	}
 
+	/**
+	 * Parses the site icon from the provided HTML
+	 *
+	 * @param string $html the HTML from the remote website at URL.
+	 * @param string $url the target website URL.
+	 * @return string the icon URI (maybe empty).
+	 */
+	private function get_icon( $html, $url ) {
+		preg_match( '|<link.*?rel="\s*[shortcut]+(?:\s+[icon]+)*\s*".*?href="(.*?)".*?\/?>|is', $html, $matches );
+
+		$icon = isset( $matches[1] ) && is_string( $matches[1] ) ? trim( $matches[1] ) : '';
+
+		// Attempt to convert relative URLs to absolute.
+		if ( ! empty( $icon ) ) {
+			$parsed = parse_url( $icon );
+			if ( empty( $parsed['host'] ) ) {
+				$icon = $url . $icon;
+			}
+		}
+
+		return $icon;
+	}
+
+	/**
+	 * Parses the meta description from the provided HTML.
+	 *
+	 * @param string $html the HTML from the remote website at URL.
+	 * @return string the meta description contents (maybe empty).
+	 */
+	private function get_description( $html ) {
+		$description = '';
+
+		$temp = tmpfile();
+
+		if ( ! $temp ) {
+			fclose( $temp ); // clean up tmp file.
+			return $description;
+		}
+
+		$path = stream_get_meta_data( $temp )['uri'];
+
+		// Write HTML.
+		fwrite( $temp, $html );
+
+		$meta = get_meta_tags( $path );
+
+		if ( empty( $meta ) ) {
+			fclose( $temp ); // clean up tmp file.
+			return $description;
+		}
+
+		$description = ! empty( $meta['description'] ) ? $meta['description'] : '';
+
+		fclose( $temp ); // clean up tmp file.
+		return $description;
```

Comment:
If a description contains HTML tags (here's [an example](https://www.theguardian.com/world/2021/may/19/eu-to-ask-uk-to-respect-citizens-rights-after-mistreatment-scandals)) in the meta `description` then 

1. Would you advise "cleaning" them?
2. If so...how? 

Often they will be in HTML entity form rather than as literal HTML strings. Presumanly because on the target website they are escaped before they are written to the meta tag.

Reply:
How about https://www.php.net/manual/en/function.html-entity-decode.php and then https://developer.wordpress.org/reference/functions/wp_strip_all_tags/.

That would ensure entities are parsed to HTML and then all the HTML is stripped leaving just the content.

Would that work?

Reply:
I guess if HTML is valid in the <meta> description tag then we shouldn't mess with it and instead expect the downstream to do the processing.

Reply:
Agreed. Let's go that way.